decidim/decidim
# Fix duplicate ActiveSupport notifications

#### :tophat: What? Why?
We have several issues opened due to fact that the same notification is being issued multiple times. Investigating the issue, i have found that sometimes Rails and spring are misbehaving. I did not find the real trigger of the issue, as it is somewhere in the rails or spring internals, and is being triggered by some evaluation of the decidim-core engine's revaluation. 

#### :pushpin: Related Issues
*Link your PR to an issue*
- Fixes #11289
- Fixes #11315

#### Testing
1. Checkout the current branch
2. in the console, go to development_app, and type: `spring stop`
3. edit your `development_app/config/spring.rb` and add the following in the `Spring.watch` array 
4. `  "../../decidim-core/lib/decidim/core/engine.rb",  "../decidim-core/lib/decidim/core/engine.rb"`
5. in the console go and type `./bin/rails c`
6. in the input type in : `ActiveSupport::Notifications.notifier.listeners_for("decidim.duplicate.events").size` and hit enter 
7. in the input type in also: `ActiveSupport::Notifications.notifier.listeners_for("decidim.events.core.welcome_notification").size` and hit enter 
8. See the both number are equal to 1. 
9. in the input type: `reload!` and hit enter 
10. repeat steps 6, 7 and 9 several times.
11. see the `ActiveSupport::Notifications.notifier.listeners_for("decidim.duplicate.events").size`  changes value
12. see the `ActiveSupport::Notifications.notifier.listeners_for("decidim.events.core.welcome_notification").size`  **does not** change it's value
13. once you consider this as a pass, make a suggestion, so i can remove 386 - 390 lines from the initializer. 

:hearts: Thank you!
